Add support for Booleans

gselzer opened the pull request:
This PR adds support for `Booleans`.

This PR is motivated by the desire to support the conversion to/from `RAI<BooleanType>` in PyImageJ.

gselzer committed:
Add support for Booleans

ctrueden commented on src/main/java/net/imglib2/img/basictypelongaccess/wrapped/WrappedBooleanLongAccess.java:
```diff
@@ -0,0 +1,76 @@
+/**
+ * TODO
+ *
+ * @author Stephan Preibisch
```
Author should be Gabriel Selzer, no? Or at least, you should be *one of* the authors.

gselzer replied:
Fixed!

gselzer committed:
Convert ByteLongAccess -> NativeBool

This is actually a much more useful way of doing things, particularly
when wrapping NumPy data

ctrueden reviewed (commented):
I presume most of this is just cut-paste-adapted from the other types, so it is likely to work the same. However, there is a question of unit tests: there aren't any. Maybe we should write some quick down-the-middle tests for all the existing types including boolean, just as an extra layer of caution?

gselzer committed:
Write a NativeBooleanUnsafe

Since they are the same size as a ByteUnsafe, let's just wrap that

hanslovsky commented on src/main/java/net/imglib2/img/basictypelongaccess/unsafe/NativeBooleanUnsafe.java:
```diff
@@ -0,0 +1,105 @@
+		wrapped.setValue(index, value ? ONE : ZERO);
+	}
+
+	public long getAddres()
+	{
+		return wrapped.getAddres();
+	}
```
Typo, might need to be fixed in `wrapped` as well. Is this a typo that already exists throughout `imglib2-unsafe`?

```suggestion
	public long getAddress()
	{
		return wrapped.getAddress();
	}
```

gselzer replied:
Yeah, I saw that too @hanslovsky, but it's all over the place. [Here](https://github.com/imglib/imglib2-unsafe/blob/5ddfbb7587fbb0e9c1ef0d3e3a13d9d6c3054f64/src/main/java/net/imglib2/img/basictypelongaccess/unsafe/DoubleUnsafe.java#L87)'s another example.

I figured this would be a breaking change to address, so I didn't fix it.

I suppose we could write `getAddress` and deprecate `getAddres` :sweat_smile:

gselzer committed:
Add myself as author

gselzer commented:
> I presume most of this is just cut-paste-adapted from the other types, so it is likely to work the same. However, there is a question of unit tests: there aren't any. Maybe we should write some quick down-the-middle tests for all the existing types including boolean, just as an extra layer of caution?

@ctrueden what would you want to test? All of this functionality assumes you have an address in memory to operate on. How do we get such an address.

That said, I'd love to test this functionality. I just don't know how.

gselzer commented:
Here's another point of discussion: in the short term, we don't actually have a use case for the vast majority of this code. I actually only want the `NativeBoolean` work right now to solve imglib/imglyb#13.

*But* we may be able to use the `BooleanType` work later, as @ctrueden outlines in [his comment](https://github.com/imglib/imglyb/issues/13#issuecomment-927480157)

ctrueden commented:
@gselzer To allocate a block of memory, I believe you can use the `long allocateMemory(long bytes)` function of `com.sun.Unsafe`, which IIUC is the equivalent of C's `malloc`. You will need to call `freeMemory` (the equivalent of C's `free`) after you are done with it, because the memory is not on the Java heap and will therefore not be garbage collected automatically.

Alternately, there is probably a way to use `ByteBuffer.allocateDirect` to make a `ByteBuffer` object and then obtain the memory address of that `ByteBuffer`.

gselzer committed:
Test Unsafe classes

gselzer committed:
Add Converter tests

gselzer commented:
@ctrueden please let me know what you think of these tests

gselzer commented:
So @ctrueden I'm actually going to remove most of the code out of this PR. We have no use for any of this `Boolean` work right now, and I don't think that `BooleanType` implementations can align with the `LongAccess` API

I'd want to support some type of `BoolLongAccessType`, with a `BoolLongAccessTypeBoolTypeConverter` (*or* the same thing but for `BitType`s), but there are problems for both `BoolType` and `BitType`:
* `BoolType`s cannot be created from any type of `Access`, be it `BooleanAccess`, `LongAccess`, etc., meaning that while we can make a `BooleanAccess` from our `BooleanLongAccessType`, we cannot make a `BoolType` that wraps the `BooleanLongAccessType`.
* `BitType`s can only be made from a `LongAccess`, which means we'd need to write a `LongLongAccessTypeBitTypeConverter`. But this then gets super confusing dimension-wise and would probably not work well.

For these reasons, I want to basically only keep the `NaiveBoolean` work. I put the other work on a different branch.

gselzer committed:
Add more support for NativeBooleanUnsafe

gselzer committed:
Move method below comment

gselzer committed:
Add Javadoc for LongAccesses

gselzer committed:
Provide some javadoc for NativeBooleanUnsafe

gselzer committed:
Add javadoc to OwningNativeBooleanUnsafe

gselzer commented:
One thing I'd like to draw attention to is the naming of `NativeBooleanUnsafe`, `OwningNativeBooleanUnsafe`, etc. This contrasts with ImgLib2's `NativeBoolType`.

I named these new classes as such since there is no definite connection between `NativeBooleanUnsafe` and `NativeBoolType`, although they interpret bits the same way.

But, this leads to a bit of confusion. For example, we can make an `ArrayImg<NativeBoolType, OwningNativeBooleanUnsafe>`, which sounds weird as the first type parameter is missing the `ean`.

So: should we name these consistently, or would that make too strong of a connection between these objects?

ctrueden commented on src/main/java/net/imglib2/img/basictypelongaccess/unsafe/NativeBooleanUnsafe.java:
```diff
@@ -0,0 +1,105 @@
+import net.imglib2.img.basictypeaccess.volatiles.VolatileBooleanAccess;
+import net.imglib2.img.basictypelongaccess.BooleanLongAccess;
+
+public class NativeBooleanUnsafe
```
This class should be named just `BooleanUnsafe`, not `NativeBooleanUnsafe`. The term `Native` in the context of ImgLib2 means "backed by a Java primitive type", and this class is not backed by `boolean[]`; it is backed by `ByteUnsafe` under the hood. It's really about providing a boolean-centric view into direct/unsafe memory access.

In this repository, classes have the word `Native` only to extend core ImgLib2's codebase from `int` indices to `long`. So we have the direct correlation:

* `AbstractNativeImg` &rarr; `AbstractNativeLongAccessImg`
* `NativeImg` &rarr; `NativeLongAccessImg`
* `NativeImgFactory` &rarr; `NativeLongAccessImgFactory`
* `AbstractNativeType` &rarr; `AbstractNativeLongAccessType`
* `NativeType` &rarr; `NativeLongAccessType`
* `NativeTypeFactory` &rarr; `NativeLongAccessTypeFactory`

Arguably, the `NativeLongAccessType` classes here should not have been named that way, because they are no longer backed by Java primitive arrays, but rather by unsafe memory addresses. (If you check the `NativeLongAccessType` class, you can see even the javadoc was copy-pasted and refers to Java primitive arrays, erroneously now.) Note that the concrete implementations such as `GenericIntLongAccessType` are again named in accordance with ImgLib2 core, with `LongAccess` sandwiched in.

So then for consistency with ImgLib2 core and boolean types, let's look what boolean types it has:

1. `BooleanType` &ndash; an interface for all boolean-based types &ndash; unfortunately named, but no changing it now
2. `BoolType` &ndash; a type for wrapping singleton `boolean` values, which unfortunately does not implement `NativeType`.
3. `NativeBoolType` &ndash; a type that *does* implement `NativeType` and can therefore be used with `ArrayImg` et al. backed by `boolean[]`. It's unfortunate that, unlike the other corresponding primitive-based `NativeType` classes, this one retained `Native` in the name and the others (e.g. `DoubleType`) did not... but it was impossible to name it just `BooleanType` because that already existed as interface (1) above, nor was the name `BoolType` available because it was already used for (2), so some differentiating factor needed to be put into the name, and `NativeBoolType` seemed the least bad/unclear.

So, this third thing `NativeBoolType` is the closest analogue to the `NativeBooleanUnsafe` here, implying this class should be named `NativeBoolUnsafe` for consistency with e.g. `DoubleType` &rarr; `DoubleUnsafe`. But unlike the situation in ImgLib2 core, then name `BooleanUnsafe` is not already taken, and it is more correct in that it is an `Unsafe`-backed access exposing the data as `boolean` values. Therefore, I vote for naming it `BooleanUnsafe`!

gselzer replied:
This all sounds great to me, I'd just appreciate some explanation added to the javadoc explaining this decision, so we remember why this was named as such!

ctrueden replied:
My writeup above was much more complex than it needed to be.

The main reason to name this class `BooleanUnsafe` is: it is the `Unsafe` access class implementing `BooleanLongAccess`. Just the same as e.g. `DoubleUnsafe` implements `DoubleLongAccess`. The `Type` implementation names don't even come into it at the `Access` level.

ctrueden reviewed (changes requested):
@gselzer Thank you very much for your work on this PR! It is mostly fabulous. :sparkler: My only concern is the naming of the `Unsafe`-backed boolean access, which I argue should be named `BooleanUnsafe`, not `NativeBooleanUnsafe`.

@gselzer If you agree, you can either push a commit renaming it, or else just give me a thumbs up and I will do it myself, before merging.

ctrueden committed:
Happy New Year 2023

ctrueden committed:
Rename NativeBooleanUnsafe to BooleanUnsafe

This class is the Unsafe analogue to ImgLib2 core's BooleanAccess
class in the net.imglib2.img.basictypeaccess package. So it should
be named BooleanUnsafe, same as e.g. int with IntAccess in core and
IntLongAccess and IntUnsafe here in imglib2-unsafe.

ctrueden committed:
Add class javadoc for Wrapped*Access classes

ctrueden committed:
Add WrappedBooleanLongAccess class

Like some others, it is not used, but for consistency it should exist.
